oauth2_http/java/com/google/auth/oauth2/GdchCredentials.java

```diff
@@ -0,0 +1,512 @@
+    tokenRequest.set("assertion", assertion);
+    UrlEncodedContent content = new UrlEncodedContent(tokenRequest);
+
+    HttpRequestFactory requestFactory = transportFactory.create().createRequestFactory();
+    HttpRequest request = requestFactory.buildPostRequest(new GenericUrl(tokenServerUri), content);
+
+    request.setParser(new JsonObjectParser(jsonFactory));
+
+    HttpResponse response;
+    String errorTemplate = "Error getting access token for GDCH service account: %s, iss: %s";
+
+    try {
+      response = request.execute();
+    } catch (HttpResponseException re) {
+      String message = String.format(errorTemplate, re.getMessage(), getServiceIdentityName());
+      throw GoogleAuthException.createWithTokenEndpointResponseException(re, message);
+    } catch (IOException e) {
+      throw GoogleAuthException.createWithTokenEndpointIOException(
+          e, String.format(errorTemplate, e.getMessage(), getServiceIdentityName()));
+    }
+
+    GenericData responseData = response.parseAs(GenericData.class);
+    String accessToken =
+        OAuth2Utils.validateString(responseData, "access_token", PARSE_ERROR_PREFIX);
+    int expiresInSeconds =
+        OAuth2Utils.validateInt32(responseData, "expires_in", PARSE_ERROR_PREFIX);
+    long expiresAtMilliseconds = clock.currentTimeMillis() + expiresInSeconds * 1000L;
+    return new AccessToken(accessToken, new Date(expiresAtMilliseconds));
+  }
+
+  /**
+   * Create a self-signed JWT for GDCH authentication flow.
+   *
+   * <p>The self-signed JWT is used to exchange access token from GDCH authentication
+   * (tokenServerUri), not for API call. It uses the serviceIdentityName as the `iss` and `sub`
+   * claim, and the tokenServerUri as the `aud` claim. The JWT is signed with the privateKey.
+   */
+  String createAssertion(JsonFactory jsonFactory, long currentTime, URI apiAudience)
+      throws IOException {
+    JsonWebSignature.Header header = new JsonWebSignature.Header();
+    header.setAlgorithm("RS256");
+    header.setType("JWT");
+    header.setKeyId(privateKeyId);
+
+    JsonWebToken.Payload payload = new JsonWebToken.Payload();
+    payload.setIssuer(getIssSubValue(projectId, serviceIdentityName));
+    payload.setSubject(getIssSubValue(projectId, serviceIdentityName));
+    payload.setIssuedAtTimeSeconds(currentTime / 1000);
+    payload.setExpirationTimeSeconds(currentTime / 1000 + this.lifetime);
+    payload.setAudience(getTokenServerUri().toString());
+
+    String assertion;
+    try {
+      payload.set("api_audience", apiAudience.toString());
+      assertion = JsonWebSignature.signUsingRsaSha256(privateKey, jsonFactory, header, payload);
+    } catch (GeneralSecurityException e) {
+      throw new IOException(
+          "Error signing service account access token request with private key.", e);
+    }
+
+    return assertion;
+  }
+
+  /** Get the issuer and subject value in the format GDCH token server required. */
```

Comment:
Let better document that this is specific to Gdch and combined parameter used for both fields

Reply:
done